### ISSUE TITLE
Add variations on the Brush tool

### ISSUE THREAD

**Issue by mriale** [mriale/PyDPainter#66]
* A segmented polygonal line, to enable grabbing non-square brushes, and avoid grabbing unwanted pixels with the square grabber.
* A lasso - drawing a freehand outline around the object
* A magic wand, grabbing an island of connected pixels surrounded by the background color.

Possible UI options:
* Right-click on the brush tool to select mode?
* Upper left / lower right subtool?
* Multiple clicks on the brush tool?

Extracted and embellished from Issue #32

**Comment by Sleuthgame** [comment]
Interesting ! Do you need DPaint-like icons for this? Subtool is very Dpaint UX but it allows only 2 tools. Multiple clicks is not how the program is expected to work. Tricky.

I fell the lasso to be useless if you can use polygonal line in zoomed part of screen or if you have magic wand.

**Comment by MetinSeven** [comment]
> Multiple clicks is not how the program is expected to work. Tricky.

I personally think Mark should not be constrained to only recreating DPaint, but in stead build upon it. PyDPainter is a freshly started tribute to DPaint, with the opportunity to enhance it.

My vote goes to multiple clicks. It's got the best balance between usability and logic. Alternatively, you could go the Photoshop way of adding a little corner icon and turning the button into a multi-button drop-down menu when pressed for a second.

**Comment by mriale** [comment]
I implemented the polygon brush tool like it is in DPaint. You have to click the brush tool a second time to use polygon mode. I also updated the tool tip text. This is currently in the `develop` branch.

**Comment by MetinSeven** [comment]
> I implemented the polygon brush tool like it is in DPaint. You have to click the brush tool a second time to use polygon mode. I also updated the tool tip text. This is currently in the `develop` branch.

That's great, thanks mark! 👍

**Comment by Sleuthgame** [comment]
Currently testing it.

Add a rapid double click or right-click to close the shape / end selection (like polygonal selection in Photoshop) ?

**Comment by Sleuthgame** [comment]
There is a bug in the interface, filled circle is now the same icon than this new tool.

**Comment by Sleuthgame** [comment]
May I suggest an icon more in the style of the brush icon ? Like this ?
![image](https://github.com/mriale/PyDPainter/assets/130657737/8f4108ed-582e-4873-b514-9ade04a5dab6)

**Comment by Sleuthgame** [comment]
![tools](https://github.com/mriale/PyDPainter/assets/130657737/c737b875-6b54-4beb-b059-49ff9faa8ab4)

**Comment by MetinSeven** [comment]
Nice, although usually a polygon tool looks like [a multi-cornered line](https://thenounproject.com/icon/polygonal-lasso-tool-177528/).

**Comment by Sleuthgame** [comment]
I can do that ! But I just tried how to. The way PydPainter handle icons in any screensize is amazing.

**Comment by Sleuthgame** [comment]
Unfortunately, complex icon like this do not reduce well in low resolution. It's unusable in 320x256 screen resolution :-( 

![tools](https://github.com/mriale/PyDPainter/assets/130657737/757d133f-be83-48d8-a58e-b64e50a01ea1)

**Comment by MetinSeven** [comment]
> Unfortunately, complex icon like this do not reduce well in low resolution. It's unusable in 320x256 screen resolution :-(
> 
> ![tools](https://user-images.githubusercontent.com/130657737/255668753-757d133f-be83-48d8-a58e-b64e50a01ea1.png)

Thanks for trying. The previous icon is fine. 👍

**Comment by Sleuthgame** [comment]
Ok, I'me some pixel away to something usuable. Maybe Mark can polish this, knowing how his toolbar resizing works.
![tools](https://github.com/mriale/PyDPainter/assets/130657737/384ac5af-265f-41c2-9552-5268436a48f7)

**Comment by mriale** [comment]
I added the dotted diamond brush to `develop` in commit 7c8d64acd255d8b71219308af73098957220c6db. I also added:
- `data/multiscale.py` - a Python program that allows you to preview images at half and quarter size. Just drag your image on it and it shows all 3 sizes. Every second it re-loads the image so you can see your changes as you save over the file with a paint program.

The method I use to scale the images down uses different pixel positions for the different sizes, so you can almost make two different images for high-res and low-res.

These images can be tiled and made semi-transparent on a modern paint program to show which pixels are used:
- `data/half_dots.png` - half-size (hi-res) pixels
- `data/quarter_dots.png` - quarter-size (low-res) pixels

I played around with the polygon for a while and gave up because nothing looked good in low-res.

**Comment by mriale** [comment]
@Sleuthgame Thanks for the idea about double-clicking to close polygons! I added double-click support to the polygon drawing tools as well as the polygon brush tool. It was always hard on DPaint to close a polygon by clicking on the first point of the polygon. That's why I added a rectangle around the first point. Double-clicking is a nice improvement and updates the user interface to more like what people expect today. This is currently in the `develop` branch.
